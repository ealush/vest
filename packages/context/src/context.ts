import { CB } from 'utilityTypes';
import {
  assign,
  defaultTo,
  invariant,
  optionalFunctionValue,
} from 'vest-utils';

export function createContext<T extends unknown>(
  defaultContextValue?: T
): CtxReturn<T> {
  let contextValue = undefined;

  return {
    use,
    run,
  };

  function use(): T {
    return defaultTo(contextValue, defaultContextValue);
  }

  function run<R>(value: T, cb: () => R): R {
    const parentContext = use();

    contextValue = value;

    const res = cb();

    contextValue = parentContext;
    return res;
  }
}

// eslint-disable-next-line max-lines-per-function
export function createCascade<T extends Record<string, unknown>>(
  init?: (ctxRef: Partial<T>, parentContext: T | void) => T | null
): CtxCascadeReturn<T> {
  let contextValue = undefined;

  return {
    bind,
    run,
    use,
    useX,
  };

  function useX(errorMessage?: string): T {
    const ctx = use();
    invariant(
      ctx,
      defaultTo(errorMessage, 'Context was used after it was closed')
    );
    return ctx;
  }

  function run<R>(ctxRef: Partial<T>, fn: () => R): R {
    const parentContext = use();

    const out = assign(
      {},
      parentContext ? parentContext : {},
      optionalFunctionValue(init, ctxRef, parentContext) ?? ctxRef
    ) as T;

    set(Object.freeze(out));

    const res = fn();

    contextValue = parentContext;
    return res;
  }

  function bind<Fn extends CB>(ctxRef: Partial<T>, fn: Fn) {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore - this one's pretty hard to get right
    const returnedFn: Fn = function (...runTimeArgs: Parameters<Fn>) {
      return run<ReturnType<Fn>>(ctxRef, function () {
        return fn(...runTimeArgs);
      });
    };

    return returnedFn;
  }

  function use() {
    return contextValue;
  }

  function set(value: T): T {
    return (contextValue = value);
  }
}

type CtxReturn<T> = {
  use: () => T;
  run: <R>(value: T, cb: () => R) => R;
};

export type CtxCascadeReturn<T> = {
  run: <R>(ctxRef: Partial<T>, fn: () => R) => R;
  bind: <Fn extends CB>(ctxRef: Partial<T>, fn: Fn) => Fn;
  use: () => T | undefined;
  useX: (errorMessage?: string) => T;
};
