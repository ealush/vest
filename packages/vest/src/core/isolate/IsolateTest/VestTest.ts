import { invariant, isPromise, optionalFunctionValue } from 'vest-utils';

import type { TIsolateTest } from 'IsolateTest';
import {
  TestAction,
  TestStateMachineAction,
  TestStatus,
  createTestStateMachine,
} from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { TFieldName, TGroupName } from 'SuiteResultTypes';

const TestStateMachine = createTestStateMachine();

export class VestTest {
  // Read

  static getData<
    F extends TFieldName = TFieldName,
    G extends TGroupName = TGroupName
  >(test: TIsolateTest<F, G>) {
    invariant(test.data);
    return test.data;
  }

  static warns(test: TIsolateTest): boolean {
    return VestTest.getData(test).severity === TestSeverity.Warning;
  }

  static isPending(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PENDING);
  }

  static isOmitted(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.OMITTED);
  }

  static isUntested(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.UNTESTED);
  }

  static isFailing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.FAILED);
  }

  static isCanceled(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.CANCELED);
  }

  static isSkipped(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.SKIPPED);
  }

  static isPassing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PASSING);
  }

  static isWarning(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.WARNING);
  }

  static hasFailures(test: TIsolateTest): boolean {
    return VestTest.isFailing(test) || VestTest.isWarning(test);
  }

  static isNonActionable(test: TIsolateTest): boolean {
    return (
      VestTest.isSkipped(test) ||
      VestTest.isOmitted(test) ||
      VestTest.isCanceled(test)
    );
  }

  static isTested(test: TIsolateTest): boolean {
    return VestTest.hasFailures(test) || VestTest.isPassing(test);
  }

  static awaitsResolution(test: TIsolateTest): boolean {
    // Is the test in a state where it can still be run, or complete running
    // and its final status is indeterminate?
    return (
      VestTest.isSkipped(test) ||
      VestTest.isUntested(test) ||
      VestTest.isPending(test)
    );
  }

  static isAsyncTest(test: TIsolateTest): boolean {
    return isPromise(VestTest.getData(test).asyncTest);
  }

  static statusEquals(test: TIsolateTest, status: TestStatus): boolean {
    return VestTest.getData(test).status === status;
  }

  // Mutate

  static setPending(test: TIsolateTest) {
    VestTest.setStatus(test, TestStatus.PENDING);
  }

  static fail(test: TIsolateTest): void {
    VestTest.setStatus(
      test,
      VestTest.warns(test) ? TestStatus.WARNING : TestStatus.FAILED
    );
  }

  static pass(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.PASSING);
  }

  static warn(test: TIsolateTest): void {
    VestTest.setData(test, current => ({
      ...current,
      severity: TestSeverity.Warning,
    }));
  }

  static setData(
    test: TIsolateTest,
    setter:
      | ((current: TIsolateTest['data']) => TIsolateTest['data'])
      | TIsolateTest['data']
  ): void {
    test.data = optionalFunctionValue(setter, VestTest.getData(test));
  }

  static skip(test: TIsolateTest, force?: boolean): void {
    // Without this force flag, the test will be marked as skipped even if it is pending.
    // This means that it will not be counted in "allIncomplete" and its done callbacks
    // will not be called, or will be called prematurely.
    // What this mostly say is that when we have a pending test for one field, and we then
    // start typing in a different field - the pending test will be canceled, which
    // is usually an unwanted behavior.
    // The only scenario in which we DO want to cancel the async test regardless
    // is when we specifically skip a test with `skipWhen`, which is handled by the
    // "force" boolean flag.
    // I am not a fan of this flag, but it gets the job done.
    VestTest.setStatus(test, TestStatus.SKIPPED, force);
  }

  static cancel(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.CANCELED);
    VestTest.getData(test).abortController.abort(TestStatus.CANCELED);
  }

  static omit(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.OMITTED);
  }

  static reset(test: TIsolateTest): void {
    VestTest.setStatus(test, TestAction.RESET);
  }

  static setStatus(
    test: TIsolateTest,
    status: TestStateMachineAction,
    payload?: any
  ): void {
    VestTest.setData(test, current => ({
      ...current,
      status: TestStateMachine.staticTransition(
        current.status,
        status,
        payload
      ),
    }));
  }
}
