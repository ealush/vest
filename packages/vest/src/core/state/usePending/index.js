import state from '..';

export default state.registerHandler(() => ({
  pending: [],
  lagging: [],
}));
