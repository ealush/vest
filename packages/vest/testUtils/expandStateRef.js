export default function expandStateRef(stateRef) {
  const state = {};
  for (const key in stateRef) {
    const [value] = stateRef[key]();
    state[key] = value;
  }
  return state;
}
