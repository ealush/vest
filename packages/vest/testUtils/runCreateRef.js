import createStateRef from 'createStateRef';
import createState from 'state';

export default state =>
  createStateRef(state ? state : createState(), {
    suiteId: 1000,
    name: 'suite_name',
  });
