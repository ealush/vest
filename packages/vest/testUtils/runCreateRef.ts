import createState from 'vast';

import createStateRef from 'createStateRef';

export default (
  state?: ReturnType<typeof createState>
): ReturnType<typeof createStateRef> =>
  createStateRef(state ? state : createState(), {
    suiteId: '1000',
  });
