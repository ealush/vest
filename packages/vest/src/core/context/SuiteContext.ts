import { createCascade } from 'context';
import { assign, TinyState, tinyState, cache, CacheApi } from 'vest-utils';
import { DynamicValue } from 'vest-utils/src/utilityTypes';

import { IsolateTest } from 'IsolateTest';
import { Modes } from 'Modes';

export const SuiteContext = createCascade<CTXType>((ctxRef, parentContext) => {
  if (parentContext) {
    return null;
  }

  return assign(
    {
      inclusion: {},
      mode: tinyState.createTinyState<Modes>(Modes.EAGER),
      suiteParams: [],
      testMemoCache,
    },
    ctxRef
  );
});

type CTXType = {
  inclusion: Record<string, DynamicValue<boolean>>;
  mode: TinyState<Modes>;
  suiteParams: any[];
  testMemoCache: CacheApi<IsolateTest>;
  currentTest?: IsolateTest;
  groupName?: string;
  skipped?: boolean;
  omitted?: boolean;
};

export function useCurrentTest(msg?: string) {
  return SuiteContext.useX(msg).currentTest;
}

export function useGroupName() {
  return SuiteContext.useX().groupName;
}

export function useInclusion() {
  return SuiteContext.useX().inclusion;
}

export function useMode() {
  return SuiteContext.useX().mode();
}

export function useSkipped() {
  return SuiteContext.useX().skipped ?? false;
}

export function useOmitted() {
  return SuiteContext.useX().omitted ?? false;
}

const testMemoCache = cache<IsolateTest>(10);

export function useTestMemoCache() {
  return SuiteContext.useX().testMemoCache;
}

export function useSuiteParams() {
  return SuiteContext.useX().suiteParams;
}
