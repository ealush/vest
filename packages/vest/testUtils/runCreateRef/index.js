import useTestCallbacks from '../../src/core/produce/useTestCallbacks';
import state from '../../src/core/state';
import useSuiteId from '../../src/core/suite/useSuiteId';
import usePending from '../../src/core/test/lib/pending/usePending';
import useTestObjects from '../../src/core/test/useTestObjects';

export default () =>
  state.createRef({
    usePending,
    useSuiteId: [useSuiteId, [1000, 'suite_name']],
    useTestCallbacks,
    useTestObjects,
  });
