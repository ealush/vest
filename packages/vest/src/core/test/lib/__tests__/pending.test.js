import _ from 'lodash';

import expandStateRef from '../../../../../testUtils/expandStateRef';
import runCreateRef from '../../../../../testUtils/runCreateRef';

import VestTest from 'VestTest';
import context from 'ctx';
import { removePending, setPending } from 'pending';
import usePending from 'usePending';

const groupName = 'group_name';

let state, stateRef;

it.ctx = (str, cb) => it(str, () => context.run({ stateRef }, cb));
beforeEach.ctx = cb => beforeEach(() => context.run({ stateRef }, cb));

describe('module: pending', () => {
  let refContent, testObject;

  const runRemovePending = testObject => {
    context.run({ stateRef }, () => removePending(testObject));
  };

  beforeEach(() => {
    stateRef = runCreateRef(state);
    refContent = _.cloneDeep(expandStateRef(stateRef));
    testObject = new VestTest({
      fieldName: 'field_1',
      statement: 'failure_message',
      testFn: jest.fn(),
    });
  });
  describe('export: removePending', () => {
    describe('When testObject it not pending or lagging', () => {
      it('Should keep state unchanged', () => {
        runRemovePending(testObject);
        expect(expandStateRef(stateRef)).toEqual(refContent);
      });
    });

    describe('When testObject is either pending or lagging', () => {
      describe('When in pending', () => {
        beforeEach.ctx(() => {
          const [, setPending] = usePending();

          setPending(state => {
            state.pending = state.pending.concat(testObject);
            return state;
          });
        });

        it.ctx('Should remove test from pending', () => {
          {
            const [pendingState] = usePending();
            expect(pendingState.pending).toContain(testObject);
          }
          runRemovePending(testObject);
          {
            const [pendingState] = usePending();
            expect(pendingState.pending).not.toContain(testObject);
          }
          expect(expandStateRef(stateRef)).toMatchSnapshot();
        });
      });
      describe('When in lagging', () => {
        beforeEach.ctx(() => {
          const [, setPending] = usePending();

          setPending(state => {
            state.lagging = state.lagging.concat(testObject);
            return state;
          });
        });

        it.ctx('Should remove test from lagging', () => {
          {
            const [pendingState] = usePending();
            expect(pendingState.lagging).toContain(testObject);
          }
          runRemovePending(testObject);
          {
            const [pendingState] = usePending();
            expect(pendingState.lagging).not.toContain(testObject);
          }
          expect(expandStateRef(stateRef)).toMatchSnapshot();
        });
      });
    });
  });

  describe('export: setPending', () => {
    let testObjects;

    beforeEach(() => {
      testObjects = Array.from(
        { length: 5 },
        (v, i) =>
          new VestTest({
            fieldName: `test_${i}`,
            statement: 'Some statement string',
            testFn: jest.fn(),
          })
      );
      testObjects[0].groupName = groupName;
    });

    it.ctx('Should set supplied test object as pending', () => {
      {
        const [pendingState] = usePending();
        expect(pendingState.pending).not.toContain(testObjects[0]);
      }
      setPending(testObjects[0]);
      {
        const [pendingState] = usePending();
        expect(pendingState.pending).toContain(testObjects[0]);
      }
      expect(expandStateRef(stateRef)).toMatchSnapshot();
    });

    describe('When a field of the same profile is in lagging array', () => {
      beforeEach.ctx(() => {
        const [, setPending] = usePending();

        setPending(state => {
          state.lagging = state.lagging.concat(
            testObjects[2], // same fieldName, group = undefined
            testObjects[0], // same fieldName, group = group_name
            testObjects[1] // same fieldName, group = undefined
          );

          return state;
        });
      });

      it.ctx('Should remove test from lagging array', () => {
        {
          const [pendingState] = usePending();
          expect(pendingState.lagging).toContain(testObjects[0]);
        }
        const added = new VestTest({
          fieldName: testObjects[0].fieldName,
          group: testObjects[0].groupName,
          statement: 'failure message',
          testFn: jest.fn(),
        });
        setPending(added);
        {
          const [pendingState] = usePending();
          expect(pendingState.pending).toContain(added);
          expect(pendingState.lagging).not.toContain(testObjects[0]);
        }
        expect(expandStateRef(stateRef)).toMatchSnapshot();
      });

      it.ctx('Should add test to pending array', () => {
        {
          const [pendingState] = usePending();
          expect(pendingState.pending).not.toContain(testObjects[0]);
        }
        setPending(testObjects[0]);
        {
          const [pendingState] = usePending();
          expect(pendingState.pending).toContain(testObjects[0]);
        }
      });

      it.ctx('Should set test as canceled', () => {
        expect(testObjects[0].canceled).toBeUndefined();
        setPending(
          new VestTest({
            fieldName: testObjects[0].fieldName,
            group: testObjects[0].groupName,
            statement: 'failure message',
            testFn: jest.fn(),
          })
        );
        expect(testObjects[0].canceled).toBe(true);
      });
    });
  });
});
