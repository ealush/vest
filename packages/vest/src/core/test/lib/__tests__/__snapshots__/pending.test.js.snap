// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module: pending export: removePending When testObject is either pending or lagging When in lagging Should remove test from lagging 1`] = `
Object {
  "pending": Object {
    "lagging": Array [],
    "pending": Array [],
  },
  "suiteId": Object {
    "id": 1000,
    "name": "suite_name",
  },
  "testCallbacks": Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Array [],
  },
  "testObjects": Array [],
}
`;

exports[`module: pending export: removePending When testObject is either pending or lagging When in pending Should remove test from pending 1`] = `
Object {
  "pending": Object {
    "lagging": Array [],
    "pending": Array [],
  },
  "suiteId": Object {
    "id": 1000,
    "name": "suite_name",
  },
  "testCallbacks": Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Array [],
  },
  "testObjects": Array [],
}
`;

exports[`module: pending export: setPending Should set supplied test object as pending 1`] = `
Object {
  "pending": Object {
    "lagging": Array [],
    "pending": Array [
      VestTest {
        "failed": false,
        "fieldName": "test_0",
        "groupName": "group_name",
        "id": "4",
        "isWarning": false,
        "skipped": false,
        "statement": "Some statement string",
        "testFn": [MockFunction],
      },
    ],
  },
  "suiteId": Object {
    "id": 1000,
    "name": "suite_name",
  },
  "testCallbacks": Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Array [],
  },
  "testObjects": Array [],
}
`;

exports[`module: pending export: setPending When a field of the same profile is in lagging array Should remove test from lagging array 1`] = `
Object {
  "pending": Object {
    "lagging": Array [
      VestTest {
        "failed": false,
        "fieldName": "test_2",
        "id": "12",
        "isWarning": false,
        "skipped": false,
        "statement": "Some statement string",
        "testFn": [MockFunction],
      },
      VestTest {
        "failed": false,
        "fieldName": "test_1",
        "id": "11",
        "isWarning": false,
        "skipped": false,
        "statement": "Some statement string",
        "testFn": [MockFunction],
      },
    ],
    "pending": Array [
      VestTest {
        "failed": false,
        "fieldName": "test_0",
        "groupName": "group_name",
        "id": "15",
        "isWarning": false,
        "skipped": false,
        "statement": "failure message",
        "testFn": [MockFunction],
      },
    ],
  },
  "suiteId": Object {
    "id": 1000,
    "name": "suite_name",
  },
  "testCallbacks": Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Array [],
  },
  "testObjects": Array [],
}
`;
